### PR TITLE
FINERACT-1724: Remove disbursement details in undo loan disbursal command

### PR DESCRIPTION
**Pull request by josehernandezfintecheandomx**
## Description

Remove the Loan Disbursement Details when the Loan Product has the flag `isDisallowExpectedDisbursements` in true as part of the Undo Loan Disbursal command

[FINERACT-1724](https://issues.apache.org/jira/browse/FINERACT-1724).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/legacy-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Review by galovics — CHANGES_REQUESTED** [review]
Tests pls.

**Comment by josehernandezfintecheandomx** [comment]
> Tests pls.

Once time the PR https://github.com/apache/fineract/pull/2519 is merged I can do more quickly this Integration Test

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 1]
```diff
@@ -869,6 +873,20 @@ public CommandProcessingResult undoLoanDisbursal(final Long loanId, final JsonCo
             final Map<String, Object> accountingBridgeData = loan.deriveAccountingBridgeData(currency.getCode(), existingTransactionIds,
                     existingReversedTransactionIds, isAccountTransfer);
             journalEntryWritePlatformService.createJournalEntriesForLoan(accountingBridgeData);
+
+            // Remove All the Disbursement Details If the Loan Product is disabled and exists one
+            if (loan.loanProduct().isDisallowExpectedDisbursements()) {
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:875]: If we have multiple disbursement details and the loan is not getting back to approved (by undoing the latest disbursement) i dont think it is a good idea to remove all of the entries.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java [hunk 2]
```diff
@@ -869,6 +873,20 @@ public CommandProcessingResult undoLoanDisbursal(final Long loanId, final JsonCo
...
+                if (!loan.getDisbursementDetails().isEmpty()) {
+                    this.loanDisbursementDetailsRepository.deleteAll(loan.getDisbursementDetails());
+                }
+            }
+
+            // Remove All the Loan Delinquency Tags If exist one
+            List<LoanDelinquencyTagHistory> loanDelinquencyTagsHistory = this.loanDelinquencyTagHistoryRepository.findByLoan(loan);
```
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:885]: Same here... if the loan is not getting back to approved state (there was multiple disbursement) we shall not remove the delinquency state

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java [hunk 3]
```diff
@@ -347,6 +350,83 @@ public void createApproveAndValidateMultiDisburseLoan() throws ParseException {
         this.editLoanDisbursementDetails();
     }
 
+    @Test
```
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java:353]: Missing test case:
1. There are multiple disbursement
1.1 Undo the 2nd one, check whether it is remains active, the principal amount is correct, the scheduling is correct

**Review by adamsaghy — CHANGES_REQUESTED** [review]
@josehernandezfintecheandomx Please check my comments!

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 2 above)
Comment (adamsaghy) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: I dont think we should remove the history entries. 
If there was any entry, it should be lifted and thats all.

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java (hunk 3 above)
Comment (adamsaghy) [inline integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java:351]: Missing test case:
1. Create and disburse loan product with "disallowExpectedDisbursements=true"
2. Undo the disbursement
3. Disburse again

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 1 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java:875]: Those now are marked as reversed and not more removed physically

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java (hunk 2 above)
Comment (josehernandezfintecheandomx) [reply fineract-provider/src/main/java/org/apache/fineract/portfolio/loanaccount/service/LoanWritePlatformServiceJpaRepositoryImpl.java]: Done

File: integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java (hunk 3 above)
Comment (josehernandezfintecheandomx) [reply integration-tests/src/test/java/org/apache/fineract/integrationtests/LoanDisbursementDetailsIntegrationTest.java:351]: Done

**Review by IOhacker — APPROVED** [review]

**Comment by IOhacker** [comment]
LGTM

**Pull request merged**